crates/sui-indexer-alt-framework/src/store.rs

```diff
@@ -0,0 +1,173 @@
+    async fn set_committer_watermark(
+        &mut self,
+        pipeline: &'static str,
+        watermark: CommitterWatermark,
+    ) -> anyhow::Result<bool>;
+
+    /// Update the reader low watermark for an existing watermark row, as long as this raises the
+    /// watermark, and updates the timestamp this update happened to the database's current time.
+    ///
+    /// Returns a boolean indicating whether the watermark was actually updated or not.
+    async fn set_reader_watermark(
+        &mut self,
+        pipeline: &'static str,
+        reader_lo: i64,
+    ) -> anyhow::Result<bool>;
+
+    /// Get the bounds for the region that the pruner still has to prune for the given `pipeline`,
+    /// along with a duration to wait before acting on this information, based on the time at which
+    /// the pruner last updated the bounds, and the configured `delay`. More specifically, this is
+    /// the result of delay + (pruner_timestamp - current_database_time)
+    ///
+    /// The pruner is allowed to prune the region between the returned `pruner_hi` (inclusive) and
+    /// `reader_lo` (exclusive) after `wait_for` milliseconds have passed since this response was
+    /// returned.
+    async fn pruner_watermark(
+        &mut self,
+        pipeline: &'static str,
+        delay: Duration,
+    ) -> anyhow::Result<Option<PrunerWatermark>>; // (pruner_hi, reader_lo, wait_for_ms)
+
+    /// Update the pruner watermark, returns true if the watermark was actually updated
+    async fn set_pruner_watermark(
+        &mut self,
+        pipeline: &'static str,
+        pruner_hi: i64,
+    ) -> anyhow::Result<bool>;
+}
+
+/// A storage-agnostic interface for managing pipeline watermarks.
+///
+/// This trait abstracts away the underlying storage implementation, providing
+/// a consistent way to connect to the database regardless of the specific
+/// storage technology being used.
+#[async_trait]
+pub trait Store: Send + Sync + 'static + Clone {
+    type Connection<'c>: DbConnection
+    where
+        Self: 'c;
+
+    async fn connect<'c>(&'c self) -> Result<Self::Connection<'c>, anyhow::Error>;
+}
+
+pub type HandlerBatch<H> = <H as SequentialHandler>::Batch;
+
+/// Extends the Store trait with transactional capabilities.
+///
+/// This trait provides methods to execute operations within a database transaction,
+/// ensuring atomicity when committing handler batches and updating watermarks.
+/// It allows for safely combining multiple database operations that must succeed
+/// or fail together.
+#[async_trait]
+pub trait TransactionalStore: Store {
+    async fn transaction<'a, R, F>(&self, f: F) -> anyhow::Result<R>
+    where
+        R: Send + 'a,
+        F: Send + 'a,
+        F: for<'r> FnOnce(
+            &'r mut Self::Connection<'_>,
+        ) -> ScopedBoxFuture<'a, 'r, anyhow::Result<R>>;
+}
+
+#[derive(Default, Debug, Clone, Copy)]
+pub struct CommitterWatermark {
+    pub epoch_hi_inclusive: i64,
```

Comment:
These should probably all be `u64` -- the fact that we use `i64` for storage-adjacent types is an implementation detail of diesel/postgres.